x/crosschain/types/tx_body_verification.go

```diff
@@ -0,0 +1,162 @@
+	}
+	if txx.ChainId().Cmp(big.NewInt(msg.ChainId)) != 0 {
+		return fmt.Errorf("invalid chain id, want evm chain id %d, got %d", txx.ChainId(), msg.ChainId)
+	}
+	switch msg.CoinType {
+	case coin.CoinType_Zeta:
+		// Inbound depositing ZETA interacts with connector contract
+		if txx.To().Hex() != chainParams.ConnectorContractAddress {
+			return fmt.Errorf("receiver is not connector contract for coin type %s", msg.CoinType)
+		}
+	case coin.CoinType_ERC20:
+		// Inbound depositing ERC20 interacts with ERC20 custody contract
+		if txx.To().Hex() != chainParams.Erc20CustodyContractAddress {
+			return fmt.Errorf("receiver is not erc20Custory contract for coin type %s", msg.CoinType)
+		}
+	case coin.CoinType_Gas:
+		// Inbound depositing Gas interacts with TSS address
+		tssAddr := eth.HexToAddress(tss.Eth)
+		if tssAddr == (eth.Address{}) {
+			return fmt.Errorf("tss address not found")
+		}
+		if txx.To().Hex() != tssAddr.Hex() {
+			return fmt.Errorf("receiver is not tssAddress contract for coin type %s", msg.CoinType)
+		}
+	default:
+		return fmt.Errorf("coin type not supported %s", msg.CoinType)
+	}
+	return nil
+}
+
+// VerifyOutTxBody verifies the tx body for a outbound tx
+func VerifyOutTxBody(msg MsgAddToOutTxTracker, txBytes []byte, tss observertypes.QueryGetTssAddressResponse) error {
+	// verify message against transaction body
+	if chains.IsEVMChain(msg.ChainId) {
+		return verifyOutTxBodyEVM(msg, txBytes, tss.Eth)
+	} else if chains.IsBitcoinChain(msg.ChainId) {
+		return verifyOutTxBodyBTC(msg, txBytes, tss.Btc)
+	}
+	return fmt.Errorf("cannot verify outTx body for chain %d", msg.ChainId)
+}
+
+// verifyOutTxBodyEVM validates the sender address, nonce, chain id and tx hash.
+func verifyOutTxBodyEVM(msg MsgAddToOutTxTracker, txBytes []byte, tssEth string) error {
+	var txx ethtypes.Transaction
+	err := txx.UnmarshalBinary(txBytes)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal transaction %s", err.Error())
+	}
+	signer := ethtypes.NewLondonSigner(txx.ChainId())
+	sender, err := ethtypes.Sender(signer, &txx)
+	if err != nil {
+		return fmt.Errorf("failed to recover sender %s", err.Error())
+	}
+	tssAddr := eth.HexToAddress(tssEth)
+	if tssAddr == (eth.Address{}) {
+		return fmt.Errorf("tss address not found")
+	}
+	if sender != tssAddr {
+		return fmt.Errorf("sender is not tss address %s", sender)
+	}
+	if txx.ChainId().Cmp(big.NewInt(msg.ChainId)) != 0 {
+		return fmt.Errorf("invalid chain id, want evm chain id %d, got %d", txx.ChainId(), msg.ChainId)
+	}
+	if txx.Nonce() != msg.Nonce {
+		return fmt.Errorf("invalid nonce, want nonce %d, got %d", txx.Nonce(), msg.Nonce)
+	}
+	if txx.Hash().Hex() != msg.TxHash {
+		return fmt.Errorf("invalid tx hash, want tx hash %s, got %s", txx.Hash().Hex(), msg.TxHash)
+	}
+	return nil
+}
+
+// verifyOutTxBodyBTC validates the SegWit sender address, nonce and chain id and tx hash
+func verifyOutTxBodyBTC(msg MsgAddToOutTxTracker, txBytes []byte, tssBtc string) error {
```

Comment:
Not required for this pr but we should add tests to improve the code coverage sections pointed out by codecov, we can can add a issue to track it

Reply:
Yeah, created this issue before: https://github.com/zeta-chain/node/issues/1994
I added a todo to track it in this file to makes it clearer

Reply:
b5578adc